packages/headless/src/features/insight-user-actions/insight-user-actions-preprocessing.test.ts

```diff
@@ -0,0 +1,645 @@
+          document: {},
+        },
+        {
+          actionType: 'CLICK',
+          timestamp: '1648744320000',
+          eventData: {},
+          searchHub: 'in-product-help',
+          document: {title: 'title', uriHash: 'caCgiG2JPzjZfS7G'},
+        },
+        {
+          actionType: 'CUSTOM',
+          timestamp: '1648744290000',
+          eventData: {
+            type: 'smartSnippetSuggestions',
+            value: 'expandSmartSnippetSuggestion',
+          },
+          searchHub: 'in-product-help',
+          document: {},
+        },
+      ];
+      const mockRawActions = [...fakeActions].slice(0, 8);
+      const mappedAndSortedActions =
+        mapAndSortActionsByMostRecent(mockRawActions);
+
+      expect(mappedAndSortedActions).toEqual(expectedMappedAndSortedActions);
+      expect(Number(mappedAndSortedActions[0].timestamp)).toBeGreaterThan(
+        Number(mappedAndSortedActions[1].timestamp)
+      );
+    });
+
+    it('should log a warning in the console if a rawUserAction cannot be parsed', () => {
+      const invalidRawUserAction = {
+        name: 'Custom',
+        value: '{ "key": "value",,,, }',
+        time: '2023-10-01T12:00:00Z',
+      };
+
+      const consoleWarnSpy = jest
+        .spyOn(console, 'warn')
+        .mockImplementation(
+          () =>
+            'Some user actions could not be parsed. Please check the raw user actions data.'
+        );
+
+      mapAndSortActionsByMostRecent([invalidRawUserAction]);
+      expect(consoleWarnSpy).toHaveBeenCalledWith(
+        'Some user actions could not be parsed. Please check the raw user actions data.'
+      );
+
+      consoleWarnSpy.mockRestore();
+    });
+  });
+
+  describe('#splitActionsIntoTimelineSessions', () => {
+    describe('when it finds a case creation session', () => {
+      it('should properly split user actions into sessions and return the timeline including current session', async () => {
+        const mockRawActions = [...fakeActions];
+        const mappedAndSortedActions =
+          mapAndSortActionsByMostRecent(mockRawActions);
+        const ticketCreationDate = JSON.stringify(caseCreationDate.getTime());
+
+        const sessions = splitActionsIntoTimelineSessions(
+          mappedAndSortedActions,
+          ticketCreationDate
+        );
+
+        expect(sessions.session?.actions.length).toEqual(6);
+        expect(Number(sessions.session?.start)).toBeGreaterThan(
+          Number(
+            sessions.precedingSessions[sessions.precedingSessions.length - 1]
+              .end
+          )
+        );
+        expect(Number(sessions.session?.end)).toBeLessThan(
+          Number(sessions.followingSessions[0].start)
+        );
+
+        expect(sessions.precedingSessions.length).toEqual(1);
+        expect(sessions.precedingSessions[0].actions.length).toEqual(3);
+
+        expect(sessions.followingSessions.length).toEqual(2);
+        expect(sessions.followingSessions[0].actions.length).toEqual(2);
```

Comment:
Instead of doing all these checks only based on length, you can create yourself the object  that represent the expected timeline, and deep equal compare it with the result of the function, I still believe snapshots could be useful  here with these tests.